net: use actual Timeout instances on Sockets

Fishrock123 opened the pull request:
This makes `net.Socket`s use actual `Timeout` objects in a `_timeout` property, rather than making the socket itself a timer and appending properties to it directly.

This should make the code generally easier to understand, and might also prevent some deopts from properties being changes on the socket itself.

It is possible this could effect performance either better or worse, but I highly doubt it would be a significant difference. I'll try to run benchmarks on it in the coming days, but if anyone else would like to that would also be appreciated.

This also exposes `timers.Timeout` so as to avoid a circular dependency with the new `lib/internal/timers.js`, but we might not want that so I should be able to refactor more of timers into the internals file if necessary. Tagging as `semver-minor` for now due to this. I could also split that into separate commits.

_Made live on https://twitch.tv/nodesource_

CI: https://ci.nodejs.org/job/node-test-pull-request/6202/

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
net, timers

cjihrig commented on lib/net.js:
```diff
@@ -199,8 +201,19 @@ function Socket(options) {
-  for (var s = this; s !== null; s = s._parent)
-    timers._unrefActive(s);
+  for (var s = this; s !== null; s = s._parent) {
+    // if (!s._timeout || s._timeout._idleTimeout === -1) continue;
```
Can you remove the commented code.

cjihrig commented on lib/net.js:
```diff
@@ -199,8 +201,19 @@ function Socket(options) {
-    timers._unrefActive(s);
+  for (var s = this; s !== null; s = s._parent) {
+    // if (!s._timeout || s._timeout._idleTimeout === -1) continue;
+    if (!s._timeout) continue;
```
Maybe a result of the commented code, but this could be

```
if (s._timeout)
  timers._unrefActive(s._timeout);
```

Fishrock123 commented:
Ah right, forgot to remove that before PRing. Updated, new CI: https://ci.nodejs.org/job/node-test-pull-request/6203/

jasnell commented on lib/net.js:
```diff
@@ -133,6 +134,7 @@ function Socket(options) {
   this._handle = null;
   this._parent = null;
   this._host = null;
+  this._timeout = null;
```
Rather than introducing yet another `_` prefixed property, I'd much prefer to use a Symbol.
Also, given that I'm following an identical pattern in the http2 work, I'd like that Symbol to be reusable (perhaps as an export off the `internal/timers.js` this PR creates

evanlucas replied:
Isn't that part of the implementation of a Timeout instance? i.e. didn't a socket already have it, it just wasn't declared in the constructor?

jasnell replied:
Not that I'm able to see. There are zero hits anywhere in the source for a `_timeout` property

evanlucas replied:
It gets added if Socket#setTimeout() is called iirc https://github.com/nodejs/node/blob/master/lib/net.js#L330

jasnell replied:
Again, not that I'm able to see: https://github.com/nodejs/node/blob/master/lib/timers.js#L296-L317

evanlucas replied:
Apologies, I was thinking of `._idleTimeout`. Disregard

Fishrock123 commented:
It should be noted that in situations where timer properties were previously available on the Socket object, they will not be after this patch. I highly doubt people rely on those though, since they may not always be there to begin with.

addaleax commented on lib/internal/timers.js:
```diff
@@ -0,0 +1,34 @@
+const timers = require('timers');
+
+// Timeout values > TIMEOUT_MAX are set to 1.
+const TIMEOUT_MAX = 2147483647; // 2^31-1
```
Can `timers.js` re-use this value by importing it from here?

lpinca commented:
Isn't this semver-major?

jasnell commented:
I think I would lean more towards semver-major on this, just to be careful

evanlucas commented:
+1 to Semver-major

Fishrock123 commented:
Theoretically `semver-major` yes. If people seem to think it should be then so be it, I don't hold strongly one way or other.

Fishrock123 commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/571/

Fishrock123 commented:
Addressed most of the remaining comments.

@jasnell I made the new property hidden using a `kTimeout` symbol. I'm slightly worried if anyone still needed to access the timeout directly on a socket after this change. Maybe we could just expose a new API if that was really necessary, or perhaps the existing APIs already cover anything that could be needed.

jasnell commented:
We *should* be safe but if someone really needed that access then adding a new API shouldn't be out of the question.

jasnell commented:
@Fishrock123 ... ping... what's the status on this one?

Fishrock123 commented:
Ugh, I'm just worried that we will end up breaking some module or something.

Would anyone be calling e.g. `timers.active()` directly on a Socket? `timers.unenroll()`? I mean I highly doubt it so probably not but who knows?

... new CI: https://ci.nodejs.org/job/node-test-pull-request/6985/

Guess we might as well land if it looks good. @nodejs/ctc if anyone else has significant reservations.

mscdex commented:
Can the new internal module use the `module.exports = {}` format for consistency?

Fishrock123 commented:
@mscdex Done, updated and rebased.

CI: https://ci.nodejs.org/job/node-test-pull-request/6987/

mscdex commented on lib/internal/timers.js:
```diff
@@ -0,0 +1,60 @@
+// Timeout values > TIMEOUT_MAX are set to 1.
+const TIMEOUT_MAX = 2147483647; // 2^31-1
+
+module.exports = {
```
This should be placed at the bottom of the file.

Fishrock123 replied:
Why? It works fine here, and other files have it like that. I don't see a good reason to change?

mscdex replied:
I believe most of the other files have it at the bottom for a couple of reasons: it's easier to find and it works for situations where you are exporting something whose value is only set later on in the file.

mscdex commented:
Out of curiosity, did you ever run any of the net benchmarks with these changes?

Fishrock123 commented:
Just tried benchmarking but it seems they take an enormously long time to run in a compare. Maybe I'll be able to run it overnight.

Fishrock123 commented:
I did get some preliminary data:

```
cat compare.csv | Rscript benchmark/compare.R
Warning message:
package ‘ggplot2’ was built under R version 3.3.2
                                               improvement confidence      p.value
 net/net-c2s-cork.js dur=5 type="buf" len=1024      1.36 %            5.631015e-02
 net/net-c2s-cork.js dur=5 type="buf" len=128       1.88 %        *** 6.371713e-07
 net/net-c2s-cork.js dur=5 type="buf" len=16        0.76 %        *** 1.640368e-05
 net/net-c2s-cork.js dur=5 type="buf" len=32        0.67 %        *** 9.700705e-06
 net/net-c2s-cork.js dur=5 type="buf" len=4         1.38 %        *** 1.083507e-06
 net/net-c2s-cork.js dur=5 type="buf" len=512       1.85 %        *** 1.068536e-05
 net/net-c2s-cork.js dur=5 type="buf" len=64        1.80 %        *** 4.898699e-14
 net/net-c2s-cork.js dur=5 type="buf" len=8         0.75 %         ** 1.144794e-03
```

I'm willing to bet that is mostly due to a pervious hidden class change.

ronkorving commented:
@Fishrock123 You currently have a conflict with test/parallel/test-net-socket-timeout.js

Fishrock123 commented:
Rebased and did a full benchmark run last night on my laptop with Wi-Fi off and no apps open. I disabled sleep but I think it still might have gone into power-savings mode...

The results seem the opposite so maybe that's new V8, or that the numbers are just not conclusive. Changes are largely insignificant anyways though.

```
                                                   improvement confidence      p.value
 net/net-c2s-cork.js dur=5 type="buf" len=1024         -1.12 %        *** 4.744149e-04
 net/net-c2s-cork.js dur=5 type="buf" len=128          -1.11 %        *** 2.230088e-05
 net/net-c2s-cork.js dur=5 type="buf" len=16           -0.44 %         ** 1.457820e-03
 net/net-c2s-cork.js dur=5 type="buf" len=32           -0.31 %            1.461147e-01
 net/net-c2s-cork.js dur=5 type="buf" len=4            -0.02 %            9.522954e-01
 net/net-c2s-cork.js dur=5 type="buf" len=512          -0.61 %          * 2.277872e-02
 net/net-c2s-cork.js dur=5 type="buf" len=64           -0.62 %          * 1.857015e-02
 net/net-c2s-cork.js dur=5 type="buf" len=8            -0.41 %          * 2.059988e-02
 net/net-c2s.js dur=5 type="asc" len=102400            -0.84 %         ** 2.056752e-03
 net/net-c2s.js dur=5 type="asc" len=16777216          -0.76 %         ** 5.147097e-03
 net/net-c2s.js dur=5 type="buf" len=102400            -0.70 %         ** 8.781217e-03
 net/net-c2s.js dur=5 type="buf" len=16777216          -0.56 %         ** 2.238673e-03
 net/net-c2s.js dur=5 type="utf" len=102400            -0.42 %         ** 2.740268e-03
 net/net-c2s.js dur=5 type="utf" len=16777216          -0.02 %            8.976417e-01
 net/net-pipe.js dur=5 type="asc" len=102400           -0.12 %            7.765516e-01
 net/net-pipe.js dur=5 type="asc" len=16777216         -0.69 %            5.750058e-01
 net/net-pipe.js dur=5 type="buf" len=102400            0.48 %            2.495812e-01
 net/net-pipe.js dur=5 type="buf" len=16777216          0.53 %            3.984729e-01
 net/net-pipe.js dur=5 type="utf" len=102400           -0.28 %            5.517754e-01
 net/net-pipe.js dur=5 type="utf" len=16777216         -0.32 %            4.417964e-01
 net/net-s2c.js dur=5 type="asc" len=102400            -0.98 %        *** 8.356080e-05
 net/net-s2c.js dur=5 type="asc" len=16777216          -0.29 %            2.561481e-01
 net/net-s2c.js dur=5 type="buf" len=102400            -0.32 %            8.830243e-02
 net/net-s2c.js dur=5 type="buf" len=16777216          -0.45 %         ** 5.161501e-03
 net/net-s2c.js dur=5 type="utf" len=102400            -0.48 %        *** 8.650774e-04
 net/net-s2c.js dur=5 type="utf" len=16777216          -0.04 %            7.943545e-01
 net/tcp-raw-c2s.js dur=5 type="asc" len=102400        -0.37 %            1.319729e-01
 net/tcp-raw-c2s.js dur=5 type="asc" len=16777216       0.26 %            5.949423e-02
 net/tcp-raw-c2s.js dur=5 type="buf" len=102400        -0.19 %            5.167163e-01
 net/tcp-raw-c2s.js dur=5 type="buf" len=16777216      -0.30 %            1.741662e-01
 net/tcp-raw-c2s.js dur=5 type="utf" len=102400        -0.54 %         ** 6.899480e-03
 net/tcp-raw-c2s.js dur=5 type="utf" len=16777216      -0.02 %            9.285933e-01
 net/tcp-raw-pipe.js dur=5 type="asc" len=102400        5.02 %            3.275301e-01
 net/tcp-raw-pipe.js dur=5 type="asc" len=16777216     -0.71 %            3.884552e-01
 net/tcp-raw-pipe.js dur=5 type="buf" len=102400       -0.03 %            9.929536e-01
 net/tcp-raw-pipe.js dur=5 type="buf" len=16777216      0.90 %            3.097973e-01
 net/tcp-raw-pipe.js dur=5 type="utf" len=102400       -5.32 %            3.845294e-01
 net/tcp-raw-pipe.js dur=5 type="utf" len=16777216     -0.75 %            4.563353e-01
 net/tcp-raw-s2c.js dur=5 type="asc" len=102400        -0.26 %            3.331969e-01
 net/tcp-raw-s2c.js dur=5 type="asc" len=16777216       0.06 %            7.040579e-01
 net/tcp-raw-s2c.js dur=5 type="buf" len=102400         0.01 %            9.702606e-01
 net/tcp-raw-s2c.js dur=5 type="buf" len=16777216      -0.35 %            6.208888e-02
 net/tcp-raw-s2c.js dur=5 type="utf" len=102400         0.06 %            7.994832e-01
 net/tcp-raw-s2c.js dur=5 type="utf" len=16777216       0.09 %            6.268283e-01
```

jasnell commented:
Needs one more rebase and CI run

Fishrock123 commented:
Rebased.

@mscdex thoughts on the benchmarks?

Fishrock123 commented:
@nodejs/ctc This is semver-major and hasn't seen that much review so... ping I guess?

addaleax commented on lib/net.js:
```diff
@@ -234,9 +238,12 @@ function Socket(options) {
 }
 util.inherits(Socket, stream.Duplex);
 
+// Refresh existing timeouts.
```
In what way does this “refresh” timeouts?

Fishrock123 replied:
If your run `active()` or `_unrefActive()` on an existing timer object it removes it from the underlying list and re-appends it to the appropriate timer pool, refreshing the timer. (Same duration but from "now" rather than when it was first scheduled.)

@addaleax Can you think of a wording that would be more helpful based on that?

addaleax replied:
Ah, thanks – I think the comment is fine, maybe `_unrefActive()` just isn’t a very descriptive name.

Fishrock123 replied:
I would agree.

addaleax commented on lib/timers.js:
```diff
@@ -23,13 +23,18 @@
 
 const TimerWrap = process.binding('timer_wrap').Timer;
 const L = require('internal/linkedlist');
+const timers = require('internal/timers');
```
Uh, could this be named something like `internalTimers`? :smile:

Fishrock123 replied:
Sure, is that a convention? I don't really mind either way though.

addaleax replied:
Grepping for `internal[A-Z]\w+ = require` in `lib/` gives 24 results, so I think it’s safe to say this is convention. :) I just noticed because at first I thought this refers to the `timers` module, and then I was confused because this *is* the source code for that module, and… you get the idea.

Fishrock123 commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/718/

Fishrock123 commented:
@nodejs/ctc I'll be merging this early next week if no other concerns are brought up.

thefourtheye commented on lib/net.js:
```diff
@@ -46,6 +47,8 @@ const errnoException = util._errnoException;
 const exceptionWithHostPort = util._exceptionWithHostPort;
 const isLegalPort = internalNet.isLegalPort;
 
+const kTimeout = require('internal/timers').kTimeout;
```
Nit: We can use object destructuring here so that both the definitions can be done in a single line.

thefourtheye commented on lib/internal/timers.js:
```diff
@@ -0,0 +1,60 @@
+    throw new TypeError('"callback" argument must be a function');
+  }
+
+  if (typeof msecs !== 'number') {
```
Isn't the `isFinite` check covers this case as well?

Fishrock123 replied:
Correct. However, this check is faster and gives a separate error message.

Fishrock123 replied:
To be clear: this is copied from `setTimeout()` in `lib/timers.js` directly.

mscdex commented on lib/internal/timers.js:
```diff
@@ -0,0 +1,60 @@
+}
+
+function setUnrefTimeout(callback, msecs) {
+  if (typeof callback !== 'function') {
```
Why is this check needed? It's being used in one place where a function is always passed.

mscdex commented on lib/internal/timers.js:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  if (typeof msecs !== 'number') {
+    throw new TypeError('"msecs" argument must be a number, was ' + msecs);
```
I know this PR is already marked as semver-major, but perhaps we could make it less so by at least using the same error message that would have been previously thrown from `.enroll()`?

jasnell replied:
... especially given that the error message will change again once this gets converted to `internal/errors`...

mscdex commented on lib/internal/timers.js:
```diff
@@ -0,0 +1,60 @@
+    throw new TypeError('"msecs" argument must be a number, was ' + msecs);
+  }
+
+  if (msecs < 0 || !isFinite(msecs)) {
```
Since we already know it's a number type, I think it may be faster to just inline the necessary checks:

```js
if (msecs < 0 || msecs !== msecs || msecs === Infinity) {
```

mscdex commented on lib/internal/timers.js:
```diff
@@ -0,0 +1,60 @@
+
+  if (msecs < 0 || !isFinite(msecs)) {
+    throw new RangeError('"msecs" argument must be ' +
+                         'a non-negative finite number, was ' + msecs);
```
I prefer 'positive, finite number'. I recall having a similar discussion in a previous PR about using 'non-negative' vs 'positive' in error messages.

As far as including the original value in the error message goes, I don't know if we've "standardized" on how it's presented, but to me the current wording is a little awkward. Perhaps it could be more explicit, like

`'... must be a positive, finite number. Saw value: ' + msecs`

or

`'... must be a positive, finite number. Received value: ' + msecs`

bnoordhuis replied:
Zero is an allowed value but is not a positive number, though.

mscdex replied:
Zero can be positive or negative.

mscdex replied:
FWIW here was the previous PR I had in mind where "non-negative" was dropped in favor of "positive": https://github.com/nodejs/node/pull/10769

Fishrock123 commented:
@mscdex Does any of this really matter, since it is internal anyways?

mscdex commented:
I think to check the performance difference of these changes, we'd need to either create a new benchmark or possibly run one of the http benchmarks because those in benchmarks/net do not exercise `socket.setTimeout()`.

mscdex commented:
@Fishrock123 What do you mean "any of this?"

mscdex commented:
Just to add, just because this function is internal does not mean suggestions should be ignored, especially because functions like `setUnrefTimeout()` are both used during runtime and are passed user values like `msecs`.

BridgeAR commented:
@Fishrock123 this needs a rebase

BridgeAR commented:
I am closing this due to long inactivity and no response.

@Fishrock123 please reopen if you would like to follow up on this.

Fishrock123 commented:
This is something I am still interested in pursuing in the very near future.

Fishrock123 commented:
Still very busy but on my short-term todo list

BridgeAR commented:
@Fishrock123 I did not fully check if all comments were addressed or if you just rebased. Could you have another look? :-)

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/11626/

Fishrock123 commented:
FUN FACT: this patch still needed all those internal checks!

...

Fishrock123 commented:
This should fail: https://ci.nodejs.org/job/node-test-pull-request/11650/

I can't get these to pass, I'm not sure why a `clientError` is emitted (this was not a problem when the patch was [first ready to land](https://github.com/nodejs/node/pull/11154#issuecomment-296227436)).

```
=== release test-http-server-keep-alive-timeout-slow-client-headers ===
Path: sequential/test-http-server-keep-alive-timeout-slow-client-headers
'HTTP/1.1 200 OK\r\nDate: Thu, 23 Nov 2017 01:12:46 GMT\r\nConnection: keep-alive\r\nContent-Length: 0\r\n\r\n'
{ Error: Parse Error
    at socketOnEnd (_http_server.js:424:20)
    at Socket.emit (events.js:164:20)
    at endReadableNT (_stream_readable.js:1062:12)
    at _combinedTickCallback (internal/process/next_tick.js:137:11)
    at process._tickCallback (internal/process/next_tick.js:179:9) bytesParsed: 0, code: 'HPE_INVALID_EOF_STATE' }
events.js:136
      throw er; // Unhandled 'error' event
      ^

Error [ERR_STREAM_WRITE_AFTER_END]: write after end
    at writeAfterEnd (_stream_writable.js:237:12)
    at Socket.Writable.write (_stream_writable.js:288:5)
    at Socket.write (net.js:739:40)
    at Timeout.setTimeout [as _onTimeout] (/Users/Jeremiah/Documents/node/test/sequential/test-http-server-keep-alive-timeout-slow-client-headers.js:41:14)
    at ontimeout (timers.js:489:11)
    at tryOnTimeout (timers.js:313:5)
    at Timer.listOnTimeout (timers.js:273:5)
Command: out/Release/node /Users/Jeremiah/Documents/node/test/sequential/test-http-server-keep-alive-timeout-slow-client-headers.js
=== release test-http-server-keep-alive-timeout-slow-server ===
Path: sequential/test-http-server-keep-alive-timeout-slow-server
events.js:136
      throw er; // Unhandled 'error' event
      ^

Error: socket hang up
    at createHangUpError (_http_client.js:330:15)
    at Socket.socketOnEnd (_http_client.js:423:23)
    at Socket.emit (events.js:164:20)
    at endReadableNT (_stream_readable.js:1062:12)
    at _combinedTickCallback (internal/process/next_tick.js:137:11)
    at process._tickCallback (internal/process/next_tick.js:179:9)
Command: out/Release/node /Users/Jeremiah/Documents/node/test/sequential/test-http-server-keep-alive-timeout-slow-server.js
```

cc @nodejs/http

Fishrock123 commented:
All checking done in this patch is necessary for consistency with other parts of the code. If you want it changed open a separate PR.

Fishrock123 commented:
Just gona open a new PR.
